lib/drivers/aws/dedicated_pool.go

```diff
@@ -0,0 +1,573 @@
+/**
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2_types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+
+	"github.com/adobe/aquarium-fish/lib/log"
+)
+
+// Custom status to set in the host for simplifying parallel ops in between the updates
+const HOST_RESERVED = "reserved"
+
+// TODO: Right now logic pinned to just one node, need to be distributed
+
+// This structure keeps the available list of hosts & state to operate on hosts management
+type dedicatedPoolWorker struct {
+	name   string
+	driver *Driver
+	record DedicatedPoolRecord
+
+	// Amount of instances per dedicated host used in capacity calculations
+	instances_per_host uint
+
+	// It's better to update active_hosts by calling updateDedicatedHosts()
+	active_hosts         map[string]ec2_types.Host
+	active_hosts_mu      sync.RWMutex
+	active_hosts_updated time.Time
+
+	// Hosts to release or scrub at specified time, used by manageHosts process
+	to_manage_at map[string]time.Time
+}
+
+// Function runs as routine and makes sure identified hosts pool fits the configuration
+func (d *Driver) newDedicatedPoolWorker(name string, record DedicatedPoolRecord) *dedicatedPoolWorker {
+	worker := &dedicatedPoolWorker{
+		name:   name,
+		driver: d,
+		record: record,
+
+		active_hosts: make(map[string]ec2_types.Host),
+		to_manage_at: make(map[string]time.Time),
+	}
+
+	// Receiving amount of instances per dedicated host
+	worker.fetchInstancesPerHost()
+
+	go worker.backgroundProcess()
+
+	log.Debugf("AWS: Created dedicated pool: %q", worker.name)
+
+	return worker
+}
+
+func (w *dedicatedPoolWorker) AvailableCapacity(instance_type string) int64 {
+	// Check if instance type fits the pool type
+	if instance_type != w.record.Type {
+		log.Warnf("AWS: dedicated %q: Incorrect pool type requested: %s", w.name, instance_type)
+		return -1
+	}
+
+	var inst_count int64
+
+	if err := w.updateDedicatedHosts(); err != nil {
+		log.Warnf("AWS: dedicated %q: Unable to update dedicated hosts list, continue with %q: %v", w.active_hosts_updated, err)
```

Comment:
reading from `active_hosts_updated` should acquire the mutex

Reply:
fenced with mutex as well